### PR TITLE
BERT example

### PR DESCRIPTION
**Pull request by conglongli**
Adding examples of BERT-style models pre-training/fine-tuning with DeepSpeed integration (e.g., ZeRO + tensor-slicing model parallelism), and how to use the example with the public Pile data (https://github.com/EleutherAI/the-pile) and reference throughput/accuracy numbers. Including API refactoring/fixes to enable the BERT case.

### CODE REVIEW

**Review by minjiaz — COMMENTED** [review]
Added suggestion after reviewing BERT examples.

File: megatron/training.py [hunk 1]
```diff
@@ -539,9 +538,13 @@ def train_step(forward_step_func, data_iterator,
             forward_backward_func = forward_backward_pipelining_without_interleaving
     else:
         forward_backward_func = forward_backward_no_pipelining
+    if args.mos or args.kd:
+        args.teacher_forward = True
```
[the remark this answers is not in the document]
Comment (conglongli) [reply megatron/training.py:549]: Thanks @minjiaz , I had applied your suggestion.

**Review by minjiaz — APPROVED** [review]
LGTM

File: megatron/model/language_model.py [hunk 2]
```diff
@@ -312,7 +312,7 @@ def __init__(self,
                  add_pooler=False,
                  pre_process=True,
                  post_process=True,
-                 num_experts=1):
+                 num_experts=[1]):
```
Comment (awan-10) [inline megatron/model/language_model.py:315]: why do we need a list here?

**Review by awan-10 — APPROVED** [review]
Looks good to me! @conglongli, thank you for refactoring the teacher model part as well for adding the comments to not change the forward_step function.

File: megatron/model/language_model.py (hunk 2 above)
Comment (conglongli) [reply megatron/model/language_model.py:315]: Because in order to support pyramid MoE we later changed args.num_experts to list type https://github.com/microsoft/Megatron-DeepSpeed/blob/b4d4a0e06577218a21abf18eb7d8088bb8ecaad3/megatron/arguments.py#L288. Some num_experts defaults in model code already changed to list (e.g, https://github.com/microsoft/Megatron-DeepSpeed/blob/b4d4a0e06577218a21abf18eb7d8088bb8ecaad3/megatron/model/transformer.py#L612), but not in this language_model.py.

Comment (awan-10) [reply megatron/model/language_model.py:315]: Got it. Thanks for catching it. I am a little surprised we didn't get any errors in running gpt.
